src/Tablebot/Handler/Permission.hs

```diff
@@ -0,0 +1,66 @@
+-- |
+-- Module      : Tablebot.Handler.Permission
+-- Description : Some internal code for handling permissions
+-- Copyright   : (c) Anna Bruce 2021
+-- License     : MIT
+-- Maintainer  : finnjkeating@gmail.com
+-- Stability   : experimental
+-- Portability : POSIX
+--
+-- This contains some functions to extract and handle privileged commands
+module Tablebot.Handler.Permission where
+
+import Control.Monad.IO.Class (liftIO)
+import Discord.Types (GuildMember, Message, RoleId, memberRoles)
+import System.Environment (lookupEnv)
+import Tablebot.Plugin.Discord (getMessageMember)
+import Tablebot.Plugin.Types
+import Tablebot.Util.Utils (isDebug)
+import Text.Read (readMaybe)
+
+data KnownRoles = KnownRoles
+  { krExec :: Maybe RoleId,
+    krModerator :: Maybe RoleId,
+    krSuperuser :: Maybe RoleId
+  }
+  deriving (Show)
+
+userHasPermission :: RequiredPermission -> UserPermission -> Bool
+userHasPermission _ (UserPerm _ _ True) = True -- Superuser always has perm
+userHasPermission None _ = True
+userHasPermission Any (UserPerm exec moderator _) = exec || moderator
+userHasPermission Exec (UserPerm exec _ _) = exec
+userHasPermission Moderator (UserPerm _ moderator _) = moderator
+userHasPermission Both (UserPerm exec moderator _) = exec && moderator
+userHasPermission _ _ = False
+
+getKnownRoles :: IO KnownRoles
+getKnownRoles = do
+  exec <- lookupEnv "EXEC_GROUP"
+  moderator <- lookupEnv "MODERATOR_GROUP"
+  superuser <- lookupEnv "SUPERUSER_GROUP"
+  return $ KnownRoles (maybeRead exec) (maybeRead moderator) (maybeRead superuser)
+  where
+    maybeRead (Just a) = readMaybe a
+    maybeRead Nothing = Nothing
+
+getMemberGroups :: Maybe GuildMember -> [RoleId]
+getMemberGroups (Just gm) = memberRoles gm
+getMemberGroups Nothing = []
+
+permsFromGroups :: Bool -> KnownRoles -> [RoleId] -> UserPermission
+permsFromGroups debug krls gps =
+  UserPerm
+    (debug || krExec krls `elemish` gps)
+    (debug || krModerator krls `elemish` gps)
+    (debug || krSuperuser krls `elemish` gps)
+  where
+    elemish (Just a) b = a `elem` b
+    elemish Nothing _ = False
```

Comment:
I think each `elemish` call is equivalent to 
```
>>= (`elem` gps)
```
i.e. instead of
```
krExec krls `elemish` gps
```
you can have
```
krExec krls >>= (`elem` gps)
```
Whether this is cleaner or not is up to you though!